### PR TITLE
Aha/storage layer1

### PR DESCRIPTION
**Pull request by andyHa**

### CODE REVIEW

File: src/main/java/sirius/biz/storage/layer1/PhysicalObjectStorage.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    private GlobalContext globalContext;
+
+    @ConfigValue("storage.layer1.spaces.default.engine")
+    private String defaultEngine;
+
+    private DerivedSpaceInfo<PhysicalStorageEngine> engines = new DerivedSpaceInfo<>(CONFIG_KEY_LAYER1_ENGINE,
+                                                                                     StorageUtils.ConfigScope.LAYER1,
+                                                                                     this::resolveStorageEngine);
+
+    private PhysicalStorageEngine resolveStorageEngine(Extension ext) {
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer1/PhysicalObjectStorage.java]: Pls longer parameter name

File: src/main/java/sirius/biz/storage/layer1/PhysicalObjectStorage.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+     *
+     * @param space the name of the space to check
+     * @return <tt>true</tt> if the space is known, <tt>false</tt> otherwise
+     */
+    public boolean isKnown(String space) {
+        return engines.contains(space);
+    }
+
+    /**
+     * Returns the storage space wit hthe given name.
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer1/PhysicalObjectStorage.java]:
```suggestion
     * Returns the storage space with the given name.
```

File: src/main/java/sirius/biz/storage/layer1/PhysicalStorageSpace.java [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+                            .withSystemErrorMessage(
+                                    "Storage Layer 1: An error occurred when uploading data to %s (%s): %s (%s)",
+                                    objectId,
+                                    name)
+                            .handle();
+        }
+    }
+
+    /**
+     * Downloads an provides the contents of the requested object.
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer1/PhysicalStorageSpace.java]:
```suggestion
     * Downloads and provides the contents of the requested object.
```

File: src/main/java/sirius/biz/storage/layer1/PhysicalStorageSpace.java [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+     *
+     * @param ctx           the request to provide a response for
+     * @param objectId      the id of the object to deliver
+     * @param fileExtension the file extension e.g. to setup a matching <tt>Content-Type</tt>
+     */
+    public void deliver(WebContext ctx, String objectId, String fileExtension) {
+        try {
+            engine.deliver(ctx,
+                           name,
+
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer1/PhysicalStorageSpace.java]: Formatting?

File: src/main/java/sirius/biz/storage/layer1/PhysicalStorageSpace.java [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+    private void handleHttpError(WebContext ctx,
+                                 String objectId,
+                                 @Nullable String fileExtension,
+                                 @Nullable String filename,
+                                 int status) {
+        PhysicalStorageSpace replicationSpace = replicationManager.getReplicationSpace(name).orElse(null);
+        if (replicationSpace == null) {
+            ctx.respondWith().error(HttpResponseStatus.valueOf(status));
+            //TODO bump stats + record event
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer1/PhysicalStorageSpace.java]: Link ticket or will this be done in one of the followup PRs?

File: src/main/java/sirius/biz/storage/layer1/PhysicalStorageSpace.java [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+                    ctx.respondWith().error(HttpResponseStatus.valueOf(nextStatus));
+                });
+            }
+        } catch (IOException e) {
+            try {
+                ctx.respondWith().error(HttpResponseStatus.INTERNAL_SERVER_ERROR);
+            } catch (Exception ex) {
+                Exceptions.ignore(ex);
+            }
+            effectiveStatus.accept(HttpResponseStatus.INTERNAL_SERVER_ERROR.code());
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer1/PhysicalStorageSpace.java]: Result of this ValueHolder is never used

File: src/main/java/sirius/biz/storage/layer1/S3StorageEngine.java [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+    private BucketName bucketName(ObjectStore store, String bucket) {
+        return store.getBucketName(bucket);
+    }
+
+    @Override
+    public void storePhysicalObject(String space,
+                                    String objectKey,
+                                    InputStream data,
+                                    long size,
+                                    Runnable invokeWhileUploading) throws IOException {
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer1/S3StorageEngine.java]: Komplett im Space handlen?

File: src/main/java/sirius/biz/storage/layer1/replication/ReplicationBackgroundLoop.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+    @Nonnull
+    @Override
+    public String getName() {
+        return "storage-layer1-replication";
+    }
+
+    @Override
+    public double maxCallFrequency() {
+        return Sirius.isStartedAsTest() ? 1d / 2 : 1d / 60;
```
Comment (sabieber) [inline src/main/java/sirius/biz/storage/layer1/replication/ReplicationBackgroundLoop.java]: why?

File: src/main/java/sirius/biz/storage/layer1/PhysicalStorageSpace.java (hunk 5 above)
Comment (andyHa) [reply src/main/java/sirius/biz/storage/layer1/PhysicalStorageSpace.java]: followup

File: src/main/java/sirius/biz/storage/layer1/replication/ReplicationBackgroundLoop.java (hunk 8 above)
Comment (andyHa) [reply src/main/java/sirius/biz/storage/layer1/replication/ReplicationBackgroundLoop.java]: to not delay all tests by a minute and we don't care too much about the additional load on the database

**Pull request merged**